common4j/src/main/com/microsoft/identity/common/java/authorities/NativeAuthCIAMAuthority.kt

```diff
@@ -0,0 +1,116 @@
+                NativeAuthCIAMAuthority {
+            // Piggy back on the existing authority creation to improve reliability.
+            val authority = Authority.getAuthorityFromAuthorityUrl(authorityUrl)
+
+            if (authority is NativeAuthCIAMAuthority) {
+                // Authority is already a NativeAuthCIAMAuthority
+                return authority as NativeAuthCIAMAuthority
+            }
+            else if (authority is CIAMAuthority) {
+                // Authority returned was a base CIAM authority
+                NativeAuthCIAMAuthority(
+                    authorityUrl = authority.authorityUri.toString(),
+                    clientId = clientId
+                )
+            }
+
+            //The authority created above is neither NativeAuthCIAMAuthority nor CIAMAuthority
+            throw ClientException(ClientException.NATIVE_AUTH_INVALID_CIAM_AUTHORITY)
+        }
+    }
+
+
+    // TODO audience, slice, flight parameters, multiple clouds supported,
+    // isAuthorityHostValidationEnabled (AzureActiveDirectoryOAuth2Configuration). Consider extending
+    // AzureActiveDirectoryAuthority
+    init {
+        mAuthorityTypeString = "AAD_NA" // AAD Native Auth
+        mAuthorityUrlString = authorityUrl
+    }
+
+    private fun createNativeAuthOAuth2Configuration(challengeTypes: List<String>?): NativeAuthOAuth2Configuration {
+       LogSession.logMethodCall(TAG, "${TAG}.createNativeAuthOAuth2Configuration")
+        return NativeAuthOAuth2Configuration(
+            authorityUrl = this.authorityURL,
+            clientId = this.clientId,
+            challengeType = getChallengeTypesWithDefault(challengeTypes)
+        )
+    }
+
+    /**
+     * Challenge types represent different authentication flows supported by the auth server and
+     * client application.
+     * e.g. "password" challenge type requires a user to supply username and password. "oob" challenge
+     * type requires user to submit a code sent to them via out of band mechanism.
+     * "redirect" is added by the SDK as a default challenge type, as the server always expects
+     * this. Duplicates are removed, and the list is then converted in a whitespace separated string
+     * (e.g. "oob password redirect")
+     */
+    private fun getChallengeTypesWithDefault(challengeTypes: List<String>?): String {
+        LogSession.logMethodCall(TAG, "${TAG}.getChallengeTypesWithDefault")
+        Logger.info(TAG, "Challenge Types passed = $challengeTypes")
+        return (challengeTypes ?: emptyList()).plus(listOf(NativeAuthConstants.GrantType.REDIRECT)).distinct().joinToString(" ")
+    }
+
+    @Throws(ClientException::class)
+    override fun createOAuth2Strategy(parameters: OAuth2StrategyParameters): NativeAuthOAuth2Strategy {
+        val config = createNativeAuthOAuth2Configuration(parameters.mChallengeTypes)
+
+        // CIAM Authorities fetch endpoints from open id configuration, communicate that to
```

Comment:
Even if the flag was set to false?

Reply:
Is this addressed?

Reply:
The flag is set to false and I have updated the comment for it.